File: solutions/python/problem1.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Tue Mar 19 05:04:39 2024

@author: touaf
"""

import hacklib

def execute():
	LIMIT = 10**14
	
	# utilisation d'une liste d'un element pour stocker la somme, car Python 
    # ne supporte ne supporte pas les variables 'non locales'
	ans = [0]
	
	# find_harshad_primes(n, digitsum, isstrong)
    # n: doit etre un nombre Harshad troncable a droite, 
    # digitsum: somme des chiffres composant n,
    # isstrong: flag indiquant si oui ou on n est un nombre premier.
    
	def find_harshad_primes(n, digitsum, isstrong):
		# Décalez d'un chiffre vers la gauche et essayez les 10 possibilités pour le chiffre le plus à droite
		m = n * 10
		s = digitsum
		for i in range(10):
			if m >= LIMIT:
				break
			if isstrong and hacklib.is_prime(m):
				ans[0] += m
			if m % s == 0:
				find_harshad_primes(m, s, hacklib.is_prime(m // s))
			m += 1
			s += 1
	
	for i in range(1, 10):  # Tous les nombres à un chiffre sont trivialement des nombres Harshad
		find_harshad_primes(i, i, False)
	return str(ans[0])


if __name__ == "__main__":
	print(execute())

File: solutions/java/problem1.java
import java.math.BigInteger;


public final class p387 {
	
	public static void main(String[] args) {
		System.out.println(new p387().run());
	}
	
	
	private static final long LIMIT = 100000000000000L;
	
	
	private BigInteger sum = BigInteger.ZERO;
	
	public String run() {
		for (int i = 1; i <= 9; i++)  // Tous les nombres à un chiffre sont trivialement des nombres Harshad
			findHarshadPrimes(i, i, false);
		return sum.toString();
	}
	
	
	// Remarque : n doit être un nombre Harshad tronquable à droite et les autres arguments sont des propriétés du nombre n.
	private void findHarshadPrimes(long n, int digitSum, boolean isStrong) {
		// Décaler d'un chiffre vers la gauche et essayer les 10 possibilités pour le chiffre le plus à droite
		long m = n * 10;
		int s = digitSum;
		for (int i = 0; i < 10 && m < LIMIT; i++, m++, s++) {
			if (isStrong && isPrime(m))
				sum = sum.add(BigInteger.valueOf(m));
			if (m % s == 0)
				findHarshadPrimes(m, s, isPrime(m / s));
		}
	}
	
	
	private static boolean isPrime(long x) {
		if (x < 0)
			throw new IllegalArgumentException("Negative number");
		if (x == 0 || x == 1)
			return false;
		for (long i = 2, end = Library.sqrt(x); i <= end; i++) {
			if (x % i == 0)
				return false;
		}
		return true;
	}
	
}
